util/conditions/experimental/merge_strategies.go

```diff
@@ -0,0 +1,408 @@
+	// create a map with all the messages and the list of objects reporting the same message.
+	messageObjMap := map[string][]string{}
+	for _, condition := range conditions {
+		if dropEmpty && condition.Message == "" {
+			continue
+		}
+
+		m := fmt.Sprintf("(%s)", condition.Status)
+		if condition.Message != "" {
+			m += fmt.Sprintf(": %s", condition.Message)
+		}
+		if _, ok := messageObjMap[m]; !ok {
+			messageObjMap[m] = []string{}
+		}
+		messageObjMap[m] = append(messageObjMap[m], klog.KRef(condition.OwnerResource.Namespace, condition.OwnerResource.Name).String())
+	}
+
+	// compute the order of messages according to the number of objects reporting the same message.
+	// Note: The message text is used as a secondary criteria to sort messages with the same number of objects.
+	messageIndex := make([]string, 0, len(messageObjMap))
+	for m := range messageObjMap {
+		messageIndex = append(messageIndex, m)
+	}
+
+	sort.SliceStable(messageIndex, func(i, j int) bool {
+		return len(messageObjMap[messageIndex[i]]) > len(messageObjMap[messageIndex[j]]) ||
+			(len(messageObjMap[messageIndex[i]]) == len(messageObjMap[messageIndex[j]]) && messageIndex[i] < messageIndex[j])
+	})
+
+	// Pick the first n messages, decrement n.
+	// For each message, add up to three objects; if more add the number of the remaining objects with the same message.
+	// Count the number of objects reporting messages not included in the above.
+	// Note: we are showing up to three objects because usually control plane has 3 machines, and we want to show all issues
+	// to control plane machines if any,
+	for i := range len(messageIndex) {
+		if *n == 0 {
+			other += len(messageObjMap[messageIndex[i]])
+			continue
+		}
+
+		m := messageIndex[i]
+		allObjects := messageObjMap[messageIndex[i]]
+		if len(allObjects) <= 3 {
+			m += fmt.Sprintf(" from %s", strings.Join(allObjects, ", "))
+		} else {
+			m += fmt.Sprintf(" from %s and %d other %s", strings.Join(allObjects[:3], ", "), len(allObjects)-3, kind)
+		}
+
+		messages = append(messages, m)
+		*n--
+	}
+
+	return messages, other
+}
+
+// getConditionsWithOwnerInfo return all the conditions from an object each one with the corresponding ConditionOwnerInfo.
+func getConditionsWithOwnerInfo(obj runtime.Object) ([]ConditionWithOwnerInfo, error) {
+	ret := make([]ConditionWithOwnerInfo, 0, 10)
+	conditions, err := GetAll(obj)
+	if err != nil {
+		return nil, err
+	}
+	for _, condition := range conditions {
+		ret = append(ret, ConditionWithOwnerInfo{
+			OwnerResource: getConditionOwnerInfo(obj),
+			Condition:     condition,
+		})
+	}
+	return ret, nil
+}
+
+// getConditionOwnerInfo return the ConditionOwnerInfo for the given object.
+func getConditionOwnerInfo(obj runtime.Object) ConditionOwnerInfo {
+	var apiVersion, kind, name, namespace string
+	kind = obj.GetObjectKind().GroupVersionKind().Kind
+	apiVersion = obj.GetObjectKind().GroupVersionKind().GroupVersion().String()
```

Comment:
apiVersion would usually be empty just like kind for typed objects. I couldn't find a usage. If we don't actually use apiVersion anywhere, can we just have a Kind field instead of embedding TypeMeta in ConditionOwnerInfo?

Reply:
Dropped, also namespace